src/games/strategy/triplea/util/TransportUtils.java

```diff
@@ -223,4 +243,74 @@ private static void loadMaxUnits(final Unit transport, final List<Unit> canBeTra
     }
   }
 
+  private static Map<Unit, List<Unit>> findTransportsThatUnitsCouldUnloadFrom(final Collection<Unit> units,
+      final Collection<Unit> transports) {
+    final List<Unit> canBeTransported = Match.getMatches(units, Matches.UnitCanBeTransported);
+    final List<Unit> canTransport = Match.getMatches(transports, Matches.UnitCanTransport);
+    final Map<Unit, List<Unit>> result = new LinkedHashMap<>();
+    for (final Unit unit : canBeTransported) {
+      final List<Unit> transportOptions = new ArrayList<Unit>();
+      for (final Unit transport : canTransport) {
+        if (containsEquivalentUnit(unit, TransportTracker.transporting(transport))) {
+          transportOptions.add(transport);
+        }
+      }
+      result.put(unit, transportOptions);
+    }
+    return result;
+  }
+
+  private static Map<Unit, List<Unit>> sortByTransportOptionsAscending(
+      final Map<Unit, List<Unit>> unitToPotentialTransports) {
+    final Map<Unit, List<Unit>> result = new LinkedHashMap<>();
+    unitToPotentialTransports.entrySet().stream()
+        .sorted((o1, o2) -> o1.getValue().size() - o2.getValue().size())
+        .forEachOrdered(e -> result.put(e.getKey(), e.getValue()));
+    return result;
+  }
+
+  private static Unit findOptimalTransportToUnloadFrom(final Unit unit,
+      final Map<Unit, List<Unit>> unitToPotentialTransports) {
```

Comment:
it's private static, and presumably does not alter static state - this makes me happy with it handling some of the dirt. But, the top level API needs a test for that to actually be acceptable IMO. With that test we can start refactoring right away whenever we want. Without that test, we would have to spend a lot of time understanding how things work, then spend time writing that test, *then* refactoring.